language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java

```diff
@@ -0,0 +1,688 @@
+                    new SnippetBlock(":", snippet + ":", snippet + ":",
+                            ItemResolverConstants.VALUE_TYPE, SnippetBlock.Kind.VALUE)));
+            if (((ReFlagsOnOffNode) nodeAtCursor.parent().parent()).rhsReFlags().isEmpty()) {
+                // Eg: re `(?i<cursor>)`
+                completionItems.add(new SnippetCompletionItem(ctx,
+                        new SnippetBlock("-", snippet + "-", snippet + "-",
+                                ItemResolverConstants.VALUE_TYPE, SnippetBlock.Kind.VALUE)));
+            }
+            addReFlags(nodeAtCursor, ctx, completionItems, snippet);
+            return completionItems;
+        } else if (isCursorAfterReOnOffFlagMinusPosition(nodeAtCursor)) {
+            // Eg: re `(?i-<cursor>)`, re `(?-<cursor>)`
+            String snippet = resolveWordForTheGivenCursorPosition(ctx, nodeFinder);
+            addReFlags(nodeAtCursor, ctx, completionItems, snippet);
+            completionItems.add(new SnippetCompletionItem(ctx, RegexSnippet.DEF_COLON_CHAR.get()));
+            return completionItems;
+        } else if (nodeAtCursor.kind() == SyntaxKind.REGEX_TEMPLATE_EXPRESSION) {
+            // Eg: re `<cursor>`, re `<cursor>()`, re `<cursor>ABC`
+            completionItems.add(new SnippetCompletionItem(ctx,
+                    Snippet.DEF_PARANTHESIS.get()));
+            completionItems.add(new SnippetCompletionItem(ctx,
+                    Snippet.DEF_SQUARE_BRACKET.get()));
+            completionItems.add(new SnippetCompletionItem(ctx,
+                    RegexpCompletionProvider.RegexSnippet.DEF_ASSERTION_START.get()));
+        }
+        if (nodeFinder.getTemplateExpressionNode() != null
+                && nodeFinder.getTemplateExpressionNode().textRange().endOffset() - 1
+                == ctx.getCursorPositionInTree()
+                && nodeAtCursor.kind() != SyntaxKind.RE_QUOTE_ESCAPE) {
+            // Eg: re `abc<cursor>`,
+
+            if (resolvedWord == null) {
+                resolvedWord = resolveWordForTheGivenCursorPosition(ctx, nodeFinder);
+            }
+            completionItems.add(new SnippetCompletionItem(ctx,
+                    new SnippetBlock("$", resolvedWord + "$", resolvedWord + "$",
+                            ItemResolverConstants.VALUE_TYPE, SnippetBlock.Kind.VALUE)));
+        }
+        return completionItems;
+    }
+
+    private static boolean isReQuantifier(NonTerminalNode nodeAtCursor) {
+        return nodeAtCursor.kind() == SyntaxKind.RE_QUANTIFIER
+                && ((ReQuantifierNode) nodeAtCursor).nonGreedyChar().isEmpty();
+    }
+
+    private static void addReFlags(NonTerminalNode nodeAtCursor, BallerinaCompletionContext ctx,
+                                   List<LSCompletionItem> completionItems, String snippet) {
+        String reFlagsSource = nodeAtCursor.parent().parent().toSourceCode();
+        for (String flag : reFlags) {
+            if (!reFlagsSource.contains(flag)) {
```

Comment:
Is it possible to avoid string operations here? And rely on syntax tree?

Reply:
No we cannot.